README.md
# Mojar

A small booster pack for Mojolicious focused on integration.
The (soft) criteria are
*   Filesystem footprint kept small.
*   Number of package dependencies kept low.
*   XS files avoided where practical.
*   Non-linux platforms (incl Strawberry Perl) supported where practical.

## Features

*   Mojar::Cache

    A bare-bones cache.  Aims to be sufficient for everyday use while providing
an easy upgrade path to CHI when better performance or richer functionality is
required.

*   Mojar::Util

    A small set of utility functions.

## See also

The real content is split out into separate distributions.

*   Mojar::Mysql

    A set of interfaces for working with MySQL databases, of most use to those
working with more than one server.

*   Mojar::Google::Analytics

    Interface for easy (unattended) downloading of reporting data.

*   Mojar::Cron

    Interface for cron calculations.

*   Mojar::BulkSms

    Interface to popular SMS sending service.


Changes
2.211 2017-06-02
  - Modify tests to accommodate Windows (log).

2.201 2017-05-22
  - Extend slurp_chomped to take path components (Util).

2.101 2017-03-27
  - Upgrade to use Mojo::File (Config).
  - Upgrade to use Mojo::File (Util).

2.051 2016-06-20
  - Avoid conditional declaration (Util).

2.041 2016-03-24
  - Small documentation improvements.

2.031 2015-02-01
  - Added slurp_chomped (Util).

2.024 2014-09-16
  - Check file exists (Config).
  - Fixed croak bug (Config).
  - Fixed portability bug (Log).

2.014 2014-06-29
  - Added as_bool (Util).
  - Added been_numeric (Util).
  - Modified spurt (Util).
  - Improved documentation (Util).
  - Fixed version (Mojar).

2.001 2014-05-21
  - Updated to accommodate new Mojo::Log (Log).
  - Added new package (Config::Ini).
  - Bumped required version of Mojolicious to 5.00.

1.071 2014-03-06
  - Config methods accept a hash of params (Config).

1.062 2014-02-10
  - Added check_exists (Util).

1.053 2014-01-20
  - Simplified pragmas when loading configuration (Config).
  - Extended documentation (Config).
  - Added new package (ClassShare).
  - Reduced version dependency to Mojolicious@3.33.

1.041 2014-01-15
  - Refactored to be a subclass instead of a fork (Log).
  - Renamed timestamp -> pattern (Log).

1.031 2014-01-14
  - Added lc_keys (Util).
  - Renamed titlecase -> unsnakecase (Util).
  - Renamed detitlecase -> snakecase (Util).
  - Reimplemented dumper to use Data::Dumper opts (Util).

1.021 2014-01-13
  - Added on_{get,set}_error (Cache).
  - Implemented max_keys (Cache).

1.016 2014-01-12
  - Tidied packaging.
  - Added sample/test config file.
  - Added copyright statements (Mojar).

1.001 2013-07-17
  - Improved packaging.
  - Extended tests (Log).
  - Fixed bug where use-time params were ignored (Log).

0.020 2012-12-31
  - Reduced quotes and indentation (Util)
  - Tweaked .travis.yml
  - Added LICENSE, MANIFEST.SKIP
  - Fixed POD bug (Mojar)
  - Added dumper (Util)
  - Removed slurp (Util)

0.014 2012-12-21
  - Added slurp, spurt (Util)


Makefile.PL
use 5.010_001;
use strict;
use warnings;

use ExtUtils::MakeMaker;

WriteMakefile(
  NAME         => 'Mojar',
  VERSION_FROM => 'lib/Mojar.pm',
  ABSTRACT     => 'Interface integration toolkit',
  AUTHOR       => 'Nic Sandfield <niczero@cpan.org>',
  LICENSE      => 'artistic_2',
  META_MERGE   => {
    no_index  => {directory => ['test']},
    requires  => {perl => 5.010001},
    resources => {
      homepage   => 'http://niczero.github.io/mojar',
      license    => 'http://www.opensource.org/licenses/artistic-license-2.0',
      repository => 'http://github.com/niczero/mojar',
      bugtracker => 'http://github.com/niczero/mojar/issues',
      x_IRC      => 'irc://irc.perl.org/#mojo'
    }
  },
  BUILD_REQUIRES => {'Test::More' => 0},
  CONFIGURE_REQUIRES => {'ExtUtils::MakeMaker' => 6.3},
  PREREQ_PM => {Mojolicious => 5},
  EXE_FILES => [],
  test => {TESTS => 'test/*.t'}
);


lib/Mojar.pm
package Mojar;
use Mojo::Base -strict;

our $VERSION = 2.211;

1;
__END__

=head1 NAME

Mojar - Integration toolkit

=head1 SYNOPSIS

  use Mojar::Util qw(snakecase unsnakecase);
  use Mojar::Cache;
  use Mojar::Log (pattern => '%R');
  use Mojar::Config;
  use Mojar::ClassShare;

=head1 DESCRIPTION

A bag of tools for integrating to various APIs.  Most of the tools are provided
in separate distributions to continue the theme of keeping your footprint small.

=head1 DISTRIBUTIONS

=over 4

=item Mojar::Mysql

Includes easy connection management, replication monitoring, and schema
analysis.

=for comment
item Mojar::Google::Analytics
Draw down your web analytics data for reporting.
item Mojar::Cron
Calculate when the next instance should run.
=end comment

=back

=head1 SUPPORT

=head2 IRC

C<nic> at C<#mojo> on C<irc.perl.org>

=head1 RATIONALE

Mojolicious is an awesome web application framework that includes many great
building blocks even for non-web development and integration.  The intention of
Mojar is to provide pluggable classes that extend that approach while getting
you closer to connecting to third-party services.

=head1 COPYRIGHT AND LICENCE

Copyright (c) 2008--2012 Sebastian Riedel.  All rights reserved.

Copyright (c) 2012--2017 Nic Sandfield.  All rights reserved.

This program is free software, you can redistribute it and/or modify it under
the terms of the Artistic License version 2.0.

All of this code is inspired by Mojolicious and the great work of Sebastian
Riedel.  In particular Config is a direct fork of its Mojolicious counterpart.


lib/Mojar/Util.pm
package Mojar::Util;
use Mojo::Base -strict;

our $VERSION = 0.351;

use B;
use Carp 'croak';
use Exporter 'import';
use Mojo::Util 'slurp';
use Scalar::Util 'reftype';
use Storable 'dclone';

our @EXPORT_OK = qw(as_bool been_numeric check_exists dumper hash_or_hashref
    loaded_path lc_keys merge slurp_chomped snakecase spurt transcribe
    unsnakecase);

# Public functions

sub as_bool {
  my ($val) = shift;
  return !! $val if been_numeric($val) or not defined $val;
  $val = lc "$val";
  return !! 1
    if $val eq '1' or $val eq 'true' or $val eq 'yes' or $val eq 'on';
  return !! undef
    if $val eq '0' or $val eq 'false' or $val eq 'no' or $val eq 'off';
  return !! $val;
}

sub dumper {
  no warnings 'once';
  require Data::Dumper;
  local $Data::Dumper::Terse = 1;
  local $Data::Dumper::Indent = 1;
  local $Data::Dumper::Quotekeys = 0;
  local $Data::Dumper::Sortkeys = 1;
  my $dump = Data::Dumper::Dumper(@_);
  $dump =~ s/\n\z//;
  return $dump;
}

sub lc_keys {
  my ($hr) = @_;
  croak q{Missing required hashref} unless reftype $hr eq 'HASH';
  %$hr = map +(lc $_ => $$hr{$_}), keys %$hr;
  return $hr;
}

sub slurp_chomped { my $a = slurp shift; () while chomp $a; $a }

sub snakecase {
  my ($string, $syllable_sep) = @_;
  $syllable_sep //= '_';
  return undef unless defined $string;
  
  my @words;
  # Absorb any leading lowercase chars
  push @words, $1 if $string =~ s/^([^A-Z]+)//;
  # Absorb each titlecase substring
  push @words, lcfirst $1 while $string =~ s/\A([A-Z][^A-Z]*)//;
  for (0 .. $#words - 1) {
    $words[$_] .= $syllable_sep unless $words[$_] =~ /[^a-z]$/;
  }
  return join '', @words;
}

sub unsnakecase {
  my ($string, $separator, $want_camelcase) = @_;
  $separator //= '_';
  return undef unless defined $string;
  
  my @words;
  # Absorb any leading separators
  push @words, $1 if $string =~ s/\A(\Q$separator\E+)//;
  # Absorb any leading component if doing camelcase
  if ($want_camelcase
      and $string =~ s/\A([^\Q$separator\E]+)\Q$separator\E?//) {
    push @words, $1;
    push @words, $1 if $string =~ s/\A(\Q$separator\E+)//;
  }
  # Absorb each substring as titlecase
  while ($string =~ s/\A([^\Q$separator\E]+)\Q$separator\E?//) {
    push @words, ucfirst lc $1;
    push @words, $1 if $string =~ s/\A(\Q$separator\E+)//;
  }
  # Fix any trailing separators
  $words[-1] .= $separator if @words && $words[-1] =~ /\A\Q$separator\E/;
  return join '', @words;
}

sub transcribe {
  my $string = shift;
  my $translator = ref $_[-1] eq 'CODE' ? pop : undef;
  return undef unless defined $string;

  my $parts = [ $string ];  # arrayref tree with strings at leaves
  my @joiners = ();  # joining string for each level
  my @level_parts = ( $parts );  # array of arrayrefs, each containing a string
  my @next_level_parts = ();  # array of arrayrefs, each containing a string
  my ($old, $new);
  while (($old, $new) = (shift, shift) and defined $new) {
    push @joiners, $new;
    foreach my $p (@level_parts) {
      # $p is arrayref containing a string
      my @components = split /\Q$old/, $p->[0], -1;
      # Modify $parts tree
      @$p = map [ $_ // '' ], @components;
      # $p is arrayref containing arrayrefs, each containing a string
      # Set up next level
      push @next_level_parts, @$p;
    }
    @level_parts = @next_level_parts;
    @next_level_parts = ();
  }
  while ($translator and my $p = shift @level_parts) {
    $p->[0] = $translator->($p->[0]);
  }

  my @traverse = ( [0, $parts] );
  while (my $next = pop @traverse) {
    my ($depth, $ref) = @$next[0,1];
    if (ref $$ref[0]) {
      if (my @deeper = grep ref($_->[0]), @$ref) {
        # Found some children not ready to be joined
        push @traverse, [$depth, $ref], map [$depth + 1, $_], @deeper;
      }
      else {
        # Children all strings => join them
        @$ref = join $joiners[$depth], map +($_->[0] //= ''), @$ref;
      }
    }
    # else string => do nothing
  }

  return $parts->[0] // '';
}

sub loaded_path {
  my ($self) = @_;
  # Try .pm
  (my $module = (ref $self // $self) .'.pm') =~ s{::}{/};
  return $INC{$module} if exists $INC{$module};

  # Try .pl
  $module =~ s{\.pm$}{.pl};
  return $INC{$module} if exists $INC{$module};

  return undef;
}

sub been_numeric {
  my $value = shift;
  # From Mojo::JSON
  return 1 if B::svref_2object(\$value)->FLAGS & (B::SVp_IOK | B::SVp_NOK)
      and 0 + $value eq $value and $value * 0 == 0;
}

sub spurt (@) {
  my $path = shift;
  my $lines = ref $_[-1] eq 'ARRAY' ? pop : \@_;
  my $count = 0;

  die qq{Can't open file "$path": $!} unless open my $file, '>', $path;
  $file->syswrite('');
  local $_;
  $file->syswrite($_), $file->syswrite($/) and ++$count for @$lines;
  close $file;
  return $count;
}

sub hash_or_hashref {
  return { @_ } if @_ % 2 == 0;  # hash
  return $_[0] if ref $_[0] eq 'HASH' or reftype $_[0] eq 'HASH';
  croak sprintf 'Hash not identified (%s)', join ',', @_;
}

sub check_exists {
  my $requireds = shift;
  my $param = hash_or_hashref(@_);
  $requireds = [$requireds] unless ref $requireds eq 'ARRAY';

  exists $param->{$_} or croak "Missing required param ($_)" for @$requireds;
  return @$param{@$requireds};
}

# Private function
sub _merge ($;$) {
  my ($left, $right) = @_;
  if (reftype $left eq 'ARRAY') {
    if (reftype $right eq 'ARRAY') {
      %{$left->[0]} = (%{$left->[0]}, %{ dclone($right->[0]) });
    }
    else {
      # $right : HASHREF
      %{$left->[0]} = (%{$left->[0]}, %{ dclone($right) });
    }
  }
  else {
    # $left : HASHREF
    if (reftype($right) eq 'ARRAY') {
      %$left = (%$left, %{ dclone($right->[0]) });
    }
    else {
      # $right : HASHREF
      %$left = (%$left, %{ dclone($right) });
    }
  }
  return $left;
}

sub merge (@);
sub merge (@) {
  # Both class & object function
#  my $class = (@_ and not ref $_[0]) ? shift : undef;
  my $class = shift unless ref $_[0];
  # defined($class) <=> class method
  return undef unless @_;
  my $left = shift;

  # $left is a ref; @right could be various things

  # If called as object method
  # 'owning' (ie leftmost) object gets modified
  # If called as class method
  # a new object is created for the result

  # It is important that the merge associates to the left
  # [ie ($a merge $b) merge $c], in contrast to Hash::Util::Simple.

  # class method => new object
  # this is done at most once per original call
  if ($class) {
    if ($left->can('clone')) {
      return merge $left->clone, @_;
    }
    elsif ($left->can('new')) {
      return merge $left->new, @_;
    }
    elsif (ref $left eq 'HASH') {
      $left = dclone($left);
    }
    else {
      croak "Unable to clone first argument\n". dumper $left;
    }
  }

  # Base case
  unless (@_) {
    return $left;
  }
  # Recurse
  elsif (@_ == 1 and ref $_[0]) {
    # object or maybe hash ref
    return _merge($left, $_[0]);
  }
  elsif (@_ > 1 and ref $_[0]) {
    # object or maybe hash ref
    my $right = shift;
    return merge _merge($left, $right), @_;
  }
  elsif (@_ > 1 and @_ % 2 == 0) {
    # assume plain hash
    return _merge($left, { @_ });
  }
  else {
    croak 'Tried to merge incompatible/non-object'. $/ . dumper(@_);
  }
}

1;
__END__

=head1 NAME

Mojar::Util - General utility functions

=head1 SYNOPSIS

  use Mojar::Util 'transcribe';

  my $replaced = transcribe $original, '_' => '-', '-' => '_';

=head1 DESCRIPTION

Miscellaneous utility functions.

=head1 FUNCTIONS

=head2 as_bool

  $boolean = as_bool($val);

Convert arbitrary scalar to a Boolean, intended to accommodate strings
equivalent to on/off, true/false, yes/no.  The following are true.

  as_bool('ON'), as_bool('true'), as_bool(42), as_bool('Yes'), as_bool('NO!')

The following are false.

  as_bool('off'), as_bool('False'), as_bool(0), as_bool('NO'), as_bool(undef)

=head2 been_numeric

  $probably_a_number = been_numeric($val);

Introspects a flag indicating whether the value has been treated as a number.
cf: L<Scalar::Util::looks_like_number>.

=head2 snakecase

  $snakecase = snakecase $titlecase;
  $snakecase = snakecase $titlecase => $separator;

Convert title-case string to snakecase.  Also converts from camelcase.

  snakecase 'iFooBar';
  # "i_foo_bar"

Rather than using an underscore, a different separator can be specified.

  snakecase 'FooBar' => '-';
  # "foo-bar"

  snakecase 'FFooBar/BazZoo' => '-';
  # "f-foo-bar/baz-zoo"

=head2 unsnakecase

  $titlecase = unsnakecase $snakecase;
  $titlecase = unsnakecase $snakecase => $separator;
  $titlecase = unsnakecase $snakecase => $separator, $want_camelcase;

Convert snake-case string to titlecase, with optional additional translations.

  unsnakecase 'foo_bar';
  # "FooBar"

  unsnakecase 'foo-bar' => '-';
  # "FooBar"

An undefined separator defaults to underscore, which is useful when you only
want to specify a camelcase result.

  unsnakecase 'foo_bar' => undef, 1;
  # "fooBar"

  unsnakecase i_foo_bar => undef, 1;
  # 'iFooBar';

There is only one level of separator; for more see C<transcribe>.

  unsnakecase 'foo-bar_baz' => '-';
  # "FooBar_baz"

Leading separators pass through.

  unsnakecase '--foo-bar' => '-';
  # "--FooBar"

As do trailing separators.

  unsnakecase '__bar_baz__';
  # "__BarBaz__"

=head2 transcribe

  $template_base = transcribe $url_path, '/' => '_';

  $controller_class =
      transcribe $url_path, '/' => '::', sub { unsnakecase $_[0] => '-' };

  $with_separators_swapped = transcribe $string, '_' => '-', '-' => '_';

Repeatedly replaces a character/string with another character/string.  Can even
swap between values, as shown in that last example.

=head2 dumper

  say dumper $object;
  print dumper($object), "\n";
  $log->debug(dumper $hashref, $arrayref, $string, $numeric);

Based on Data::Dumper it is simply a tidied (post-processed) version.  It is
argument-greedy and if passed more than one argument will wrap them in an
arrayref and then later strip away that dummy layer.  In the resulting string,
"TOP" refers to the top-most (single, possibly invisible) entity.

This is intended to be clear and succinct to support error messages and debug
statements.  It is not suitable for serialising entities because it does not try
to maintain round trip stability.  (ie Don't try to evaluate its output.)

=head2 hash_or_hashref

  $hashref = hash_or_hashref({ A => 1, B => 2 });
  $hashref = hash_or_hashref($object);  # $object if hashref-based
  $hashref = hash_or_hashref(A => 1, B => 2);
  $hashref = hash_or_hashref();  # {}

Takes care of those cases where you want to handle both hashes and hashrefs.
Always gives a hashref if it can, otherwise dies.

=head2 check_exists

  sub something {
    my ($self, %param) = @_;
    check_exists [qw(dbh log)], %param;

  sub something2 {
    my $self = shift;
    my ($dbh, $log) = check_exists [qw(dbh log)], @_;

  sub something3 {
    my ($self, $param) = @_;
    my ($dbh) = check_exists 'dbh', $param;

  package MyClass;
  use Mojar::Util ();
  sub exists = {
    my $keys = ref $_[0] eq 'ARRAY' ? $_[0] : [ @_ ];
    return Mojar::Util::exists($keys, $self);
  }

Checks that required parameters have been passed.  Takes a string or arrayref of
strings that are required keys, and a hash or hashref of parameters.  Throws an
exception if one or more strings do not exist as keys in the parameters;
otherwise returns the list of parameter values.

=head1 SEE ALSO

L<Mojo::Util>, L<String::Util>, L<Data::Dump>.


lib/Mojar/Cache.pm
package Mojar::Cache;
use Mojo::Base -base;

# Attributes

has namespace => 'main';
has on_get_error => sub { sub { require Carp; Carp::croak @_ } };
has on_set_error => sub { sub { require Carp; Carp::croak @_ } };

has slots => sub { [] };
has 'max_keys';

# Public methods

sub new {
  my ($proto, %param) = @_;
  my $self = $proto->SUPER::new(%param);
  $self->{store} = {};
  return $self;
}

# Getting and setting

sub get {
  my ($self, $key) = @_;
  my $value;
  eval {
    $value = $self->{store}{$self->namespace}{$key};
    1;
  }
  or do {
    my $e = $@ // 'Failed get';
    $self->on_get_error->($e);
  };
  return $value;
}

sub set {
  my ($self, $key, $value) = @_;
  eval {
    my $slots = $self->slots;
    unless ($self->is_valid($key)) {
      push @$slots, $key;
      while ($self->max_keys and scalar @$slots > $self->max_keys) {
        # Too many in the bed
        $self->remove(shift @$slots);
      }
    }
    $self->{store}{$self->namespace}{$key} = $value;
    1;
  }
  or do {
    my $e = $@ // 'Failed set';
    $self->on_set_error->($e);
  };
  return $self;
}

sub compute {
  my ($self, $key, $code) = @_;
  my $cache = $self->{store}{$self->namespace};
  return $cache->{$key} if exists $cache->{$key};

  my $value = $code->($key);
  $self->set($key => $value);
  return $value;
}

# Removing

sub remove {
  my ($self, $key) = @_;
  delete $self->{store}{$self->namespace}{$key};
  return $self;
}

# Inspecting keys

sub is_valid {
  my ($self, $key) = @_;
  return exists $self->{store}{$self->namespace}{$key};
}

# Atomic operations

sub append {
  my ($self, $key, $further_text) = @_;
  $self->{store}{$self->namespace}{$key} .= $further_text;
  return $self;
}

# Namespace operations

sub clear {
  my $self = shift;
  $self->{store}{$self->namespace} = {};
  return $self;
}

sub get_keys { keys %{ $_[0]{store}{$_[0]->namespace} } }

# Multiple key/value operations

sub get_multi_arrayref {
  my ($self, $keys_ref) = @_;
  my $cache = $self->{store}{$self->namespace};
  return [ map $cache->{$_}, @$keys_ref ];
}

sub get_multi_hashref {
  my ($self, $keys_ref) = @_;
  my $cache = $self->{store}{$self->namespace};
  return { map $_ => $cache->{$_}, @$keys_ref };
}

sub set_multi {
  my ($self, $hashref) = @_;
  while (my ($k, $v) = each %$hashref) {
    $self->set($k => $v);
  }
  return $self;
}

sub remove_multi {
  my ($self, $keys_ref) = @_;
  $self->remove($_) foreach @$keys_ref;
  return $self;
}

sub to_hashref { $_[0]{store}{$_[0]->namespace} || {} }

1;
__END__

=head1 NAME

Mojar::Cache - Bare-bones in-memory cache

=head1 SYNOPSIS

  use Mojar::Cache;

  my $cache = Mojar::Cache->new;
  $cache->set(foo => 'bar');
  my $foo = $cache->get('foo');

=head1 DESCRIPTION

A minimalist cache intended to be easily upgradeable to L<CHI>.

=head1 ATTRIBUTES

L<Mojar::Cache> implements the following attributes.

=head2 C<namespace>

  $cache->namespace('Admin');
  $namespace = $cache->namespace;

Namespace for the cache, defaults to C<main>.

#=head2 C<on_get_error>
#
#  my $cb = $cache->on_get_error;
#  $cache = $cache->on_get_error(sub { ... });
#
#Callback triggered by C<get> exception, defaults to a null sub.
#
#=head2 C<on_set_error>
#
#  my $cb = $cache->on_set_error;
#  $cache = $cache->on_set_error(sub { ... });
#
#Callback triggered by C<set> exception, defaults to a null sub.

=head1 METHODS

=head2 C<get>

  my $value = $cache->get('foo');

Get cached value.

=head2 C<set>

  $cache = $cache->set(foo => 'bar');

Set cache value.

=head1 RATIONALE

Mojo::Cache is beautifully simple but sometimes too barebones; CHI is the proper
way to do serious cacheing but cannot be done on a small footprint.  I wanted
minor extensions to the former with an upgrade path to the latter.

=head1 SEE ALSO

L<Mojo::Cache>, L<CHI>.

=cut


lib/Mojar/Config.pm
package Mojar::Config;
use Mojo::Base -base;

our $VERSION = 0.042;
# Adapted from Mojolicious::Plugin::Config (3.57)

use Carp 'croak';
use Mojo::Util qw(decode slurp);

sub load {
  my ($self, $file, %param) = @_;
  $param{log}->debug(sprintf 'Reading config file (%s)', $file) if $param{log};
  croak "Failed to find file ($file)" unless -f $file or -l $file;
  my $content = decode 'UTF-8', slurp $file;
  return $self->parse(\$content, %param);
}

sub parse {
  my ($self, $content_ref, %param) = @_;

  # Run Perl code
  my $config = eval sprintf '%s%s%s', 'package Mojar::Config::Sandbox;',
      'use strict;', $$content_ref;
  croak qq{Failed to load configuration from file: $@} if not $config and $@;
  croak qq{Config file did not return a hash reference.\n}
    unless ref $config eq 'HASH';
  $param{log}->debug('Config content successfully read') if $param{log};

  return $config;
}

1;
__END__

=head1 NAME

Mojar::Config - Perl-ish configuration utility for standalone code

=head1 SYNOPSIS

  use Mojar::Config;
  my $config = Mojar::Config->load('cfg/defaults.conf');
  say $config->{redis}{ip};

=head1 DESCRIPTION

A simple configuration file reader for a configuration written as a perl hash.

=head1 USAGE

  # cfg/defaults.conf
  {
    debug => undef,
    expiration => 60 * 60 * 10,
    graffiti => sprintf('%s and %s', 'love', 'peace'),
    secrets => [qw(where wild things roam)],
    redis => {
      ip => '192.168.1.1',
      port => 6379
    }
  }

The contents are evaluated, so compuatations are valid.

=head1 METHODS

=head2 load

  $hashref = Mojar::Config->load('path/to/file.conf');
  $hashref = Mojar::Config->load('path/to/file.conf', log => $log);

Loads a perl-ish configuration from the given file path.  In normal usage, this
is the only method required.  The result is a plain (unblessed) hashref.

=head2 parse

  $content = '{ testing => 2 * 2 }';
  $config = Mojar::Config->parse(\$content);
  say $config->{testing};

Does the actual parsing of the configuration, being passed a ref to the
configuration text.

=head1 DEBUGGING

Both methods accept a Mojar::Log/Mojo::Log object in their parameters.  If
passed a debug-level logger, some debugging statements become available.

  my $log = Mojar::Log->new(level => 'debug', path => '/tmp/stuff.log');
  my $config = Mojar::Config->new->load('/etc/stuff.conf', log => $log);

=head1 SEE ALSO

This is a fork of L<Mojolicious::Plugin::Config> (v3.57) that can be used
independently of having a Mojolicious app.  So if your code is for a Mojolicious
app, it makes sense to use the upstream module instead.


lib/Mojar/ClassShare.pm
package Mojar::ClassShare;
use strict;
use warnings;

our $VERSION = 0.001;
# Adapted from Mojo::Base::attr

sub import {
  my $class = shift;
  if (@_ and shift eq 'have') {
    my $caller = caller;
    no strict 'refs';
    *{"${caller}::have"} = sub { class_attr($caller, @_) };
  }
}

# Private function

sub croak ($) { require Carp; goto &Carp::croak; }

# Public method

sub class_attr {
  my ($class, $attrs, $default) = @_;
  return unless ($class = ref $class || $class) && $attrs;

  croak 'Default has to be a code reference or constant value'
    if ref $default && ref $default ne 'CODE';

  for my $attr (@{ref $attrs eq 'ARRAY' ? $attrs : [$attrs]}) {
    croak qq{Attribute "$attr" invalid} unless $attr =~ /^[a-zA-Z_]\w*$/;

    # Header (check arguments)
    my $code = "package $class;\nsub $attr {\n  no strict 'refs';\n";
    $code .= "  if (\@_ == 1) {\n";

    # No default value (return value)
    unless (defined $default) { $code .= "    return \$_[0]{'$attr'};" }

    # Default value
    else {

      # Return value
      $code .= "    return \$_[0]{'$attr'} if exists \$_[0]{'$attr'};\n";

      # Return default value
      $code .= "    return \$_[0]{'$attr'} = ";
      $code .= ref $default eq 'CODE' ? '$default->($_[0]);' : '$default;';
    }

    # Store value
    $code .= "\n  }\n  \$_[0]{'$attr'} = \$_[1];\n";

    # Footer (return invocant)
    $code .= "  \$_[0];\n}";

    warn "-- Attribute $attr in $class\n$code\n\n" if $ENV{MOJO_BASE_DEBUG};
    croak "Mojo::Base error: $@" unless eval "$code;1";
  }
}

1;
__END__

=head1 NAME

Mojar::ClassShare - Attributes for both class and objects

=head1 SYNOPSIS

  package Somewhere::MyClass;
  use Mojar::ClassShare 'have';
  use Mojar::Config;

  have config => sub { Mojar::Config->load('/etc/myapp.conf') };
  $target_ip = Somewhere::MyClass->config->{target_ip};

=head1 DESCRIPTION

Provides attributes that can be used as class attributes or object attributes.
When the invocant is the class name, the class's shared storage is used.  When
the invocant is an object, that object's individual storage is used.

The storage of each object and the class's share are separate from each other.

  have 'value';
  ...
  $value_to_use = $object->value // MyClass->value;  # using share as a default

(It is unusual for class defaults to be dynamic, so that example is just for
illustration.)

=head1 CAVEATS

Sometimes, perhaps during debugging, you want to know where your class
attributes are really stored.  The answer is under a hashref that has the same
name as the class, but actually resides in the namespace above.  In the synopsis
example, any class attributes are stored in a hashref called $MyClass in the
Somewhere namespace.  Personally I don't see this as a problem if I own that
namespace, but be aware that this isn't an approach everyone is going to like.

I hope it is already clear, but the shared attributes are only accessible via
the classname.  When accessed via an object, the object's individual storage is
used (so the behaviour is the same as Mojo::Base::attr).

The share is only shared within its perl interpreter.  In any environment where
there are multiple instances of perl (eg hypnotoad, apache, job queues) each
process maintains its own class shares.  If your class share is fairly static,
that usually just means you need to intialise it each time an interpreter starts
up.  But if your objects update the class share, in a multi-process setup those
updates won't be shared with objects in other processes.

=head1 SEE ALSO

This module is only useful if you want cheap class attributes.  If you only want
object attributes, please go to the upstream module L<Mojo::Base>, from which
this one is forked.  If on the other hand you think you want more sophisticated
class shares, CPAN has many modules with richer implementations.

=cut


lib/Mojar/Log.pm
package Mojar::Log;
use Mojo::Base 'Mojo::Log';

our $VERSION = 1.001;

use Mojo::Util 'encode';
use POSIX 'strftime';

sub import {
  my $pkg = shift;
  # Mixin
  if (@_) {
    my @args = @_;
    my $default = sub { Mojar::Log->new(@args) };
    eval sprintf
    'package %s; no strict q{refs}; sub log { $_[0]{log} //= $default->() }; 1',
        scalar caller
      or die 'Failed to create Mojar::Log mixin';
  }
}

# Attributes

has pattern => '%Y%m%d %X';

has format => sub {
  my $self = shift; weaken $self; sub {
    my ($time, $level, @lines) = @_;
    encode 'UTF-8', sprintf '%s[%s] %s', strftime($self->pattern,
        localtime($time)), $level, join "\n", @lines, ''
  }
};

1;
__END__

=head1 NAME

Mojar::Log - Simple logger

=head1 SYNOPSIS

  use Mojar::Log;

  # Log to STDERR without timestamps
  my $log = Mojar::Log->new(pattern => undef);

  # Customise log file location and timestamp pattern
  my $log = Mojar::Log->new(path => '/tmp/abc.log', pattern => '[%F %X] ');

  # Log messages
  $log->debug('Hmmm?');
  $log->info(q{We're charging per character});
  $log->warn('Uh-oh!');
  $log->error(q{You won't believe this});
  $log->fatal('OH NOES!');

=head1 DESCRIPTION

L<Mojar::Log> is a simple logger extending L<Mojo::Log>.  The additions are the
ability to set defaults at 'use' time (via a mixin) and a more ISO-ish timestamp
with support for customising that pattern.  (Mojo::Log now let's you customise
the format by passing a coderef/closure to the format attribute.)

=head1 USAGE

Standard usage is by creating a Log object as in the SYNOPSIS.  For shared code,
large codebases, or long-running processes, that is the only recommended usage.
In standalone classes and scripts it's often more convenient to set the
parameters via an implicit mixin.  In a standalone script this could be done as:

  use Mojar::Log (
    path => '/var/log/something.log',
    level => 'info',
    pattern => '%F %X...'
  );
  main->log->debug('Go!');

If the same 'use' is employed in a class then it could be done as:

  package MyClass;
  use Mojar::Log (...);

  my $o = MyClass->new(...);
  $o->log->debug('Go!');  # object method
  MyClass->log->debug('Go!');  # class method

Employing use-time parameters entails the implicit creation of a method (mixin)
and a hash entry.  The mixin works by assigning a Log object to the caller's
'log' hash key, so only works with objects implemented as hashrefs.  In other
words, only use the convenience of the mixin if that's what you were going to do
anyway.  [Using the package name to hold a hashref is considered hacky and makes
debugging a little trickier; creating a log object per caller object can be
considered sub-optimal.  If either of those trouble you, please re-read the
intro to this section then return to the SYNOPSIS.]

Thanks to L<Mojolicious> being so versatile, you can even use Mojar::Log in
those projects, taking advantage of the introduced 'pattern' attribute.

  package MyApp;
  use Mojo::Base 'Mojolicious';
  use Mojar::Log;
  sub startup {
    my $self = shift;
    $self->log(Mojar::Log->new(pattern => '...', path => '...'));
  }

And in a Lite app:

  use Mojolicious::Lite;
  use Mojar::Log;
  app->log(Mojar::Log->new(pattern => '...', path => '...'));

After which you can use the usual Mojolicious 'log' method.

=head1 ATTRIBUTES

L<Mojar::Log> inherits its attributes from L<Mojo::Log> and adds the following.

=head2 pattern

  $pattern = $log->pattern;
  $log     = $log->pattern('%Y%m%d %H:%M:%S');

Pattern to use for the timestamp.  The default pattern (above) is a fairly
minimal 17 characters.  The timestamp can be disabled altogether by setting it
to 'undef'.

  $log->pattern(undef)->info('Timeless!');  # $log then uses no timestamp

See L<Time::CTime> for the full list of specifiers, but a few common choices are
the following.

  $log->pattern('[%FT%X] ');  # ISO 8601 timestamp with secs and brackets
  $log->pattern('[%F %R] ');  # ISO 8601 timestamp omitting 'T' and secs
  $log->pattern('%y%m%d%H%M%S');  # A more minimal 12 chars

On a linux system, you can test your pattern by calling C<date>.

  date +'%Y.%m.%d %H.%M.%S'

=head1 LEGACY NEEDS

The current version of this requires at least v5 of Mojolicious.  If you need to
work with an older version of Mojolicious, consider using v1.062.

  cpanm Mojar@1.062

=head1 SEE ALSO

L<Mojo::Log>, the parent class which provides the majority of documentation.
